### ISSUE TITLE
ArrayIndexOutOfBounds | Analysis    &    NPE | AnalysisUtils (Duplicate)

### ISSUE THREAD

**Issue by SlimeDog** [Rsl1122/Plan-PlayerAnalytics#61]
Spigot 1.11.2 latest, Plan 2.8.4, sqlite DB
 
Server running empty, no player connected.

```[11:04:53 WARN]: [Plan] Plugin Plan v2.8.4 generated an exception while executing task 36976
java.lang.ArrayIndexOutOfBoundsException
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_121]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:1.8.0_121]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_121]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_121]
        at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:598) ~[?:1.8.0_121]
        at java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:677) ~[?:1.8.0_121]
        at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:735) ~[?:1.8.0_121]
        at java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:160) ~[?:1.8.0_121]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:174) ~[?:1.8.0_121]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233) ~[?:1.8.0_121]
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418) ~[?:1.8.0_121]
        at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:583) ~[?:1.8.0_121]
        at main.java.com.djrapitops.plan.utilities.Analysis.analyze(Analysis.java:136) ~[?:?]
        at main.java.com.djrapitops.plan.utilities.Analysis$1.run(Analysis.java:62) ~[?:?]
        at org.bukkit.craftbukkit.v1_11_R1.scheduler.CraftTask.run(CraftTask.java:71) ~[spigot-1.11.2.jar-2017-03-19-0743:git-Spigot-96235ab-dfee3d9]
        at org.bukkit.craftbukkit.v1_11_R1.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:52) [spigot-1.11.2.jar-2017-03-19-0743:git-Spigot-96235ab-dfee3d9]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_121]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_121]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_121]
Caused by: java.lang.ArrayIndexOutOfBoundsException
        at java.util.ArrayList.addAll(ArrayList.java:580) ~[?:1.8.0_121]
        at main.java.com.djrapitops.plan.utilities.Analysis.lambda$analyze$5(Analysis.java:177) ~[?:?]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184) ~[?:1.8.0_121]
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374) ~[?:1.8.0_121]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[?:1.8.0_121]
        at java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:291) ~[?:1.8.0_121]
        at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731) ~[?:1.8.0_121]
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) ~[?:1.8.0_121]
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) ~[?:1.8.0_121]
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692) ~[?:1.8.0_121]
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157) ~[?:1.8.0_121]
```

and another, an hour later.

```[12:04:55 WARN]: [Plan] Plugin Plan v2.8.4 generated an exception while executing task 48667
java.lang.NullPointerException
        at main.java.com.djrapitops.plan.utilities.AnalysisUtils.lambda$transformSessionDataToLengths$3(AnalysisUtils.java:71) ~[?:?]
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374) ~[?:1.8.0_121]
        at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580) ~[?:1.8.0_121]
        at main.java.com.djrapitops.plan.utilities.AnalysisUtils.transformSessionDataToLengths(AnalysisUtils.java:70) ~[?:?]
        at main.java.com.djrapitops.plan.utilities.Analysis.analyze(Analysis.java:191) ~[?:?]
        at main.java.com.djrapitops.plan.utilities.Analysis$1.run(Analysis.java:62) ~[?:?]
        at org.bukkit.craftbukkit.v1_11_R1.scheduler.CraftTask.run(CraftTask.java:71) ~[spigot-1.11.2.jar-2017-03-19-0743:git-Spigot-96235ab-dfee3d9]
        at org.bukkit.craftbukkit.v1_11_R1.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:52) [spigot-1.11.2.jar-2017-03-19-0743:git-Spigot-96235ab-dfee3d9]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_121]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_121]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_121]```

In general, too many NPEs. Please scan the code, and try/catch object assignments to protect against NPEs. Finding them one-by-one through issue tickets is a drag.

**Comment by Rsl1122** [comment]
Yeah I've definitely learned that I can't assume that variables just work. Too bad I didn't have it in my head in December when I was doing the rewrite.
Had I been using TDD (Test-driven development), most of the issues would have been avoided.
It is only recently that I got to writing tests, and so far the tests have caught 15 Exceptions that were fixed. Biggest problem is that there is so much code (with probably close to 100 variables) to test now that writing the tests takes a long time.
I don't have any tests written for Analysis yet.
